WebCrypto X25519 algorithm docs

hamishwillee opened the pull request:
Add some live examples of using the X25519 algorithm


Related docs work can be tracked in #35280

hamishwillee committed:
SubtleCrypto.generateKey() example

hamishwillee committed:
deriveKey() example

hamishwillee committed:
Fix bug

hamishwillee committed:
Fix bug pt 2

hamishwillee committed:
deriveBits() example

hamishwillee commented:
@wbamberg Given you just reviewed the other webcrypto algorithm, I've added you as a possible reviewer to this (should you so wish. Since we already added X25519 last time, this adds some live examples in the places it can be used.

hamishwillee committed:
Keys created in try block

hamishwillee committed:
Expand on X25519 benefits

hamishwillee commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
```diff
@@ -104,7 +104,11 @@ ECDH is specified in [RFC 6090](https://datatracker.ietf.org/doc/html/rfc6090).
 X25519 is a key agreement algorithm like ECDH, but built on the [Curve25519](https://en.wikipedia.org/wiki/Curve25519) elliptic curve, which is part of the Edwards-Curve Digital Signature Algorithm (EdDSA) family of algorithms defined in {{rfc("8032")}}.
-The algorithm is specified in {{rfc("7748")}}.
+
+The Curve25519 algorithms are widely used in cryptography, and are considered to be some of the most efficient/fast available.
+By comparison with the NIST (National Institute of Standards and Technology) curve key exchange algorithms used with ECDH, Curve25519 is simpler to implement, and its non-governmental origin means that the decisions behind its design choices are transparent and open.
```
Figured it is worth adding a bit more background. This summary is synthesized from the bug report https://bugzilla.mozilla.org/show_bug.cgi?id=1904836#c4 and wikipedia. Essentially it is faster  than the NIST algorithms, and because it is easier to implement, less likely to be broken by a poor implementation.

The parameters used for the NIST algorithms were also selected through an opaque "government" process and are less trusted by some people that the Curve25519 algorithms, for which the decision making process is perceived to be fully open.

wbamberg commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
````diff
@@ -187,6 +191,266 @@ async function agreeSharedSecretKey() {
+```
+
+```css hidden
+input {
+  display: block;
+}
+#log {
+  height: 150px;
+  width: 90%;
+  white-space: pre-wrap; /* wrap pre blocks */
+  overflow-wrap: break-word; /* break on words */
+  overflow-y: auto;
+  padding: 0.5rem;
+  border: 1px solid black;
````
This could use some margins to group elements: it's very hard to read as is.

wbamberg commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
````diff
@@ -187,6 +191,266 @@ async function agreeSharedSecretKey() {
+```
+
+The code below shows how we use `deriveKey()`.
+This takes the algorithm and public key of the remote party, the private key of the local party, they algorithm of the derived key, and the "usages" for the new key, and creates a derived key.
````
This has a couple of issues. It's a bit hard to read because of the multiple "and": "they algorithm of the derived key, and the "usages" for the new key, and creates a derived key" so it would be better split into two sentences, one for params and one for return value. Also, though, it reads like an exhaustive list, but then omits the `false` parameter. Also an a exhaustive list seem like just rehashing the API reference up the page. Maybe it would be better if we described the parameters in more concrete terms:

```suggestion
We pass in the remote party's X25519 public key, the local party's X25519 private key, and specify that the derived key should be an AES-GCM key. We also set the derived key to be non-extractable, and suitable for encryption and decryption.

The function returns the derived AES-GCM key.
```

hamishwillee replied:
> Also an a exhaustive list seem like just rehashing the API reference up the page. Maybe it would be better if we described the parameters in more concrete terms:

Absolutely. Much better.

wbamberg commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
````diff
@@ -187,6 +191,266 @@ async function agreeSharedSecretKey() {
+}
+```
+
+Next we define the functions that Alice will use to UTF-8 encode and then encrypt her plaintext message, and that Bob will use to decrypt and then decode the message.
````
```suggestion
Next we define the functions that Alice will use to {{glossary("UTF-8")}} encode and then encrypt her plaintext message, and that Bob will use to decrypt and then decode the message.
```

wbamberg commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
````diff
@@ -187,6 +191,266 @@ async function agreeSharedSecretKey() {
+```
+
+Next we define the functions that Alice will use to UTF-8 encode and then encrypt her plaintext message, and that Bob will use to decrypt and then decode the message.
+The both take the shared key of the caller, an "initialization vector", and the text to be encrypted or decrypted.
+
+The [AES-GCM](/en-US/docs/Web/API/SubtleCrypto/encrypt#aes-gcm) algorithm requires that the initialization vector passed to the `encrypt()` and `decrypt()` methods is the same.
+Note however that while it has to be unique, it does not have to be secret, and can be shared with the remote party along with the public keys (in this case we'll just make it directly available).
+For more information see {{domxref("AesGcmParams")}}, which describes the object passed to the methods.
````
```suggestion
They both take as arguments the shared AES key, an [initialization vector](/en-US/docs/Web/API/AesGcmParams#iv), and the text to be encrypted or decrypted.

The same initialization vector must be used for encryption and decryption, but it does not need to be secret, so usually it is sent alongside the encrypted message. In this case, though, since we're not actually sending a message, we just make it directly available.
```

I don't think the quotes on "initialization vector" are helpful here, and we do have a good place to link. Also we would _not_ generally shared the IV "along with the public keys", because typically we would only share the public key once, then generate one (or more) secret keys, and use the shared to to encrypt more than one message. whereas IVs must not be reused for more than one message.

hamishwillee replied:
Yes, this was rubbish. Thank you.

wbamberg commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
```diff
@@ -187,6 +191,266 @@ async function agreeSharedSecretKey() {
+
+The `agreeSharedSecretKey()` function below is called on loading to generate pairs and shared keys for Alice and Bob.
+It also adds a click handler for the "Encrypt" button that will trigger encryption and then decryption of the text defined in the first `<input>`.
+Note that alll the code is inside a `try...catch` handler, to ensure that we can log the case where key generation fails because the X25519 algorithm is not supported.
```
```suggestion
Note that all the code is inside a `try...catch` handler, to ensure that we can log the case where key generation fails because the X25519 algorithm is not supported.
```

wbamberg commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
```diff
@@ -187,6 +191,266 @@ async function agreeSharedSecretKey() {
+#### Result
+
+Press the Encrypt button to encrypt the text in the top `<input>` element, displaying the encrypted cypertext and decrypted cyphertext in the following two elements.
+The log area at the bottom provide information about the keys that are generated by the code.
```
Could we have a blank line added after each "Encrypt" click? Otherwise it gets really hard to read the log output if you click more than once:

```
Created Alices's key pair: (algorithm: {"name":"X25519"}, usages: deriveKey)
Created Bob's key pair: (algorithm: {"name":"X25519"}, usages: deriveKey)
alicesSecretKey: secret (algorithm: {"name":"AES-GCM","length":256}, usages: encrypt,decrypt), 
bobsSecretKey: secret (algorithm: {"name":"AES-GCM","length":256}, usages: encrypt,decrypt), 
Plaintext: The lion roars near dawn
encryptedMessage: 203,106,123,194,253...[40 bytes total]
decryptedCypertext: The lion roars near dawn
Plaintext: The lion roars near dawn
encryptedMessage: 203,106,123,194,253...[40 bytes total]
decryptedCypertext: The lion roars near dawn
```

wbamberg commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
```diff
@@ -187,6 +191,266 @@ async function agreeSharedSecretKey() {
+    const encryptedInput = document.querySelector("#encrypted");
+    const decryptedInput = document.querySelector("#decrypted");
+
+    // Define the initialization vector that will be used for encrypting and decrypting the message.
+    const initializationVector = window.crypto.getRandomValues(
+      new Uint8Array(8),
+    );
```
IV must not be used for more than one message. So this must be created anew in each encrypt call, as ECDH does: https://github.com/mdn/dom-examples/blob/08dadaa1f2447d7412c9c8488a304f28b53dfadf/web-crypto/derive-key/ecdh.js#L26.

hamishwillee replied:
Fixed by moving it into the click handler. Thank you.

wbamberg reviewed (commented):
All right, here are some comments on the derivekey part.

hamishwillee commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
```diff
@@ -187,6 +191,266 @@ async function agreeSharedSecretKey() {
+
+#### HTML
+
+First we define an HTML {{HTMLElement("input")}} for user to enter the plaintext message that Alice will send.
```
> (Although this is a bit weird because usually Alice is is considered "the user".)

Good point. I think I will make that a bit less weird like this:

```suggestion
First we define an HTML {{HTMLElement("input")}} you to enter the plaintext message that "Alice" will send.
```

hamishwillee commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
```diff
@@ -187,6 +191,266 @@ async function agreeSharedSecretKey() {
+#### HTML
+
+First we define an HTML {{HTMLElement("input")}} for user to enter the plaintext message that Alice will send.
+This is followed by another two elements for displaying the cyphertext after Alice has encypted the plain text with her derived private key, and for displaying the text after Bob has decrypted it with his derived private key.
```
Thank you - I was assuming "private key" could be used for any key that isn't to be shared. Do you think it easonable to update https://developer.mozilla.org/en-US/docs/Glossary/Key with the definition and link from this page?

wbamberg replied:
You mean something like:

"In [symmetric-key cryptography](https://developer.mozilla.org/en-US/docs/Glossary/Symmetric-key_cryptography), the same key is used for both encryption and decryption."

->

"In [symmetric-key cryptography](https://developer.mozilla.org/en-US/docs/Glossary/Symmetric-key_cryptography), the same key is used for both encryption and decryptio. This key is known as the _secret key_."

I would also have a para break before the bit about public-key crypto.

> link from this page?

I don't know about that. It's a very generic glossary page and we are deep in the weeds here. If we want this para to be absolutely clear, it would be better to so say "AES key" rather than "secret key".

hamishwillee replied:
Yes, that is exactly what I mean, though I used the words:

> This key is referred to as the _shared key_, or the _secret key_.

Feel free to directly subedit i you don't like that.

hamishwillee committed:
Apply suggestions from code review - first lot.

Co-authored-by: wbamberg <will@bootbonnet.ca>

wbamberg commented on files/en-us/web/api/subtlecrypto/derivebits/index.md:
```diff
@@ -142,6 +142,143 @@ Promise.all([generateAlicesKeyPair, generateBobsKeyPair]).then((values) => {
 
+### X25519
+
+In this live example Alice and Bob each generate an X25519 key pair.
```
```suggestion
In this example Alice and Bob each generate an X25519 key pair.
```

(just a suggestion. I doubt most readers know what "live" means here, and it ought to be obvious they can interact with it.)

wbamberg commented on files/en-us/web/api/subtlecrypto/derivebits/index.md:
```diff
@@ -142,6 +142,143 @@ Promise.all([generateAlicesKeyPair, generateBobsKeyPair]).then((values) => {
+
+#### HTML
+
+The HTML is just a button that an be pressed to generate all the keys and the shared key.
```
This doesn't seem right. Two buttons, one derives a shared secret with the two keys pairs, the other generates new key pairs.

hamishwillee replied:
Grrr. Changed it half way through.

```suggestion
The HTML is defines two buttons.
The "Change keys" button can be pressed to generate new key pairs for Alice and Bob.
The "Derive bits" button can be pressed to derive a shared secret with the current set of key pairs.
```

hamishwillee committed:
Update files/en-us/web/api/subtlecrypto/derivekey/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

wbamberg commented on files/en-us/web/api/subtlecrypto/derivebits/index.md:
```diff
@@ -142,6 +142,143 @@ Promise.all([generateAlicesKeyPair, generateBobsKeyPair]).then((values) => {
+  // Generate 2 ECDH key pairs: one for Alice and one for Bob
+  // In more normal usage, they would generate their key pairs
+  // separately and exchange public keys securely
+  changeKeys();
+  log("Keys changed");
```
The async is a bit messed up here. If you try this in a browser that doesn't support X25519 (like Safari) you get:

```
Keys changed
NotSupportedError: The operation is not supported.
``` 

...because `changeKeys()` returns a Promise, and we log "Keys changed" before the promise fulfills.

I think it might work better if you take the try...catch out of `changeKeys()`, an put it here instead, like:

```js
try {
  await changeKeys();
  log(Keys changed");
}
catch(e) {
  log(e);
}
```

I didn't test this though.

hamishwillee replied:
Good catch - this always reports key change even if on fail. 

I fixed this by moving the log() into the success-path of the existing try-catch.
That way it is after an `await` and only gets run when the promise has fulfilled.

Chrome doesn't support this algorithms and displays:

> NotSupportedError: Failed to execute 'generateKey' on 'SubtleCrypto': Algorithm: Unrecognized name

wbamberg commented on files/en-us/web/api/subtlecrypto/derivebits/index.md:
````diff
@@ -142,6 +142,143 @@ Promise.all([generateAlicesKeyPair, generateBobsKeyPair]).then((values) => {
+});
+```
+
+The code below adds a handler function that is invoked every time the "Derive-bits" button is pressed.
````
```suggestion
The code below adds a handler function that is invoked every time the "Derive bits" button is pressed.
```

wbamberg commented on files/en-us/web/api/subtlecrypto/derivebits/index.md:
```diff
@@ -142,6 +142,143 @@ Promise.all([generateAlicesKeyPair, generateBobsKeyPair]).then((values) => {
+
+#### Result
+
+Press the "Derive-bits" button to generate and log a shared secret from Bob and Alices keys.
```
```suggestion
Press the "Derive bits" button to generate and log a shared secret from Bob and Alice's keys.
```

wbamberg commented on files/en-us/web/api/subtlecrypto/derivebits/index.md:
```diff
@@ -142,6 +142,143 @@ Promise.all([generateAlicesKeyPair, generateBobsKeyPair]).then((values) => {
+#### Result
+
+Press the "Derive-bits" button to generate and log a shared secret from Bob and Alices keys.
+Press the "Change keys" button to change the keys used by both parties.
```
```suggestion
Press the "Change keys" button to change the X25519 keys used by both parties.
```

Maybe? To be clear about which keys we are changing.

wbamberg commented on files/en-us/web/api/subtlecrypto/generatekey/index.md:
```diff
@@ -225,6 +225,90 @@ The information about the created keys is logged below (or an error string if th
+
+This code generates an X25519 public and private key pair that can be used in {{domxref("SubtleCrypto.deriveKey()")}} to create a shared key, or in {{domxref("SubtleCrypto.deriveBits()")}} to create a shared secret.
+
+The code is almost exactly the same as that for the Ed25519 algorithm in the previous section, except that it uses a different algorithm name, and the allowed `keyUsages` are different.
```
Same comment as before about referring to the previous example. Just a suggestion though.

hamishwillee replied:
It is reasonable - you can create "too much context"

wbamberg commented on files/en-us/web/api/subtlecrypto/generatekey/index.md:
```diff
@@ -225,6 +225,90 @@ The information about the created keys is logged below (or an error string if th
+Code for generating a key pair using the `X25519` algorithm and logging the information in each key is shown below.
+Note that the code is run in a `try..catch` block because not all browsers support this algorithm.
+
+The JavaScript first gets the `#sign-button` and `#message` {{HTMLElement("input")}} elements, then adds a listener for the `click` event on the button.
+The event handler clears the log and runs the other operations passing the content of the `<input>` element.
```
This doesn't seem to be correct. Copy/paste error?

hamishwillee replied:
Good catch.
```suggestion
The JavaScript first gets the `#run-button` and `#log` {{HTMLElement("input")}} elements, then adds a listener for the `click` event on the button.
The event handler clears the log and runs the other operations passing the content of the `<input>` element.
```

hamishwillee committed:
Layout only changes

hamishwillee committed:
Fix up layout on example rendering

hamishwillee commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
````diff
@@ -187,6 +191,266 @@ async function agreeSharedSecretKey() {
 }
 ```
 
+### X25519 derived shared key
````
@wbamberg Because this is a live example and above the algorithm description has the same name.

I'll fix this by giving all of the examples a more expansive heading. Linter giving me trouble with this right now (it thinks because I rename one of the headings I have broken the links to it)

hamishwillee replied:
Fixed. Thanks for the `--no-verify` tip.

hamishwillee committed:
Update files/en-us/web/api/subtlecrypto/derivebits/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee commented on files/en-us/web/api/subtlecrypto/generatekey/index.md:
```diff
@@ -225,6 +225,90 @@ The information about the created keys is logged below (or an error string if th
 
+### X25519 key generation
+
+This code generates an X25519 public and private key pair that can be used in {{domxref("SubtleCrypto.deriveKey()")}} to create a shared key, or in {{domxref("SubtleCrypto.deriveBits()")}} to create a shared secret.
+
+The code is almost exactly the same as that for the Ed25519 algorithm in the previous section, except that it uses a different algorithm name, and the allowed `keyUsages` are different.
+
```
```suggestion
This code generates an X25519 public and private key pair that can be used in {{domxref("SubtleCrypto.deriveKey()")}} to create a shared key, or in {{domxref("SubtleCrypto.deriveBits()")}} to create a shared secret.

```

hamishwillee commented on files/en-us/web/api/subtlecrypto/derivebits/index.md:
```diff
@@ -142,6 +142,143 @@ Promise.all([generateAlicesKeyPair, generateBobsKeyPair]).then((values) => {
+const deriveBitsButton = document.querySelector("#buttonDeriveKeys");
+
+deriveBitsButton.addEventListener("click", async () => {
+  // Generate 2 ECDH key pairs: one for Alice and one for Bob
```
```suggestion
  // Generate 2 X25519 key pairs: one for Alice and one for Bob
```

hamishwillee committed:
Apply suggestions from code review - pt2

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee committed:
deriveBits() log inside the await

hamishwillee committed:
deriveKey - iv generate on every click

hamishwillee committed:
fix one field

hamishwillee committed:
Fix up example sidebars

hamishwillee commented:
@wbamberg Rebasing and waiting for checks to go through but I think I have addressed your issues. Thank you for the thorough review. 

Note, BCD has not yet updated on MDN, so algorithms not displays as I would expect for this.

OnkarRuikar commented:
> Still got those URL issues, but if tried locally the links work.

The URL checker has a bug. You'll have to merge with the failing workflow until the bug is fixed.

wbamberg commented on files/en-us/web/api/subtlecrypto/derivebits/index.md:
```diff
@@ -142,6 +142,146 @@ Promise.all([generateAlicesKeyPair, generateBobsKeyPair]).then((values) => {
+
+const changeKeysButton = document.querySelector("#buttonChangeKeys");
+
+changeKeysButton.addEventListener("click", async () => {
+  // Generate 2 X25519 key pairs: one for Alice and one for Bob
+  // In more normal usage, they would generate their key pairs
+  // separately and exchange public keys securely
+  changeKeys();
+});
```
I think at this point you can just do this:

```suggestion
// Generate 2 X25519 key pairs: one for Alice and one for Bob
// In more normal usage, they would generate their key pairs
// separately and exchange public keys securely
changeKeysButton.addEventListener("click", changeKeys);
```

can't you?

hamishwillee replied:
You're right!

wbamberg commented on files/en-us/web/api/subtlecrypto/generatekey/index.md:
```diff
@@ -225,6 +225,88 @@ The information about the created keys is logged below (or an error string if th
+Note that the code is run in a `try..catch` block because not all browsers support this algorithm.
+
+The JavaScript first gets the `#run-button` and `#log` {{HTMLElement("input")}} elements, then adds a listener for the `click` event on the button.
+The event handler clears the log and runs the other operations passing the content of the `<input>` element.
```
"passing the content of the `<input>` element" still seems wrong?

```suggestion
The event handler clears the log, generates an X25519 key pair, and logs some of its properties.
```

?

hamishwillee replied:
Yeah, that is much better.

wbamberg reviewed (commented):
Thanks for the updates!

Just a few more odds and ends.

OnkarRuikar commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
```diff
@@ -187,10 +194,277 @@ async function agreeSharedSecretKey() {
+
+#### Result
+
+Press the Encrypt button to encrypt the text in the top `<input>` element, displaying the encrypted cypertext and decrypted cyphertext in the following two elements.
```
`cypher` is [british](https://www.merriam-webster.com/dictionary/cypher).

hamishwillee replied:
Also Australian :-(

OnkarRuikar commented:
>> Still got those URL issues, but if tried locally the links work.
>
> The URL checker has a bug. You'll have to merge with the failing workflow until the bug is fixed.

I've pulled the fix: https://github.com/mdn/content/pull/35577. Can we wait for the fix to get merged to test the GitHub workflow? Then, we can rebase the branch and see if it works in Github.

wbamberg commented:
> > > Still got those URL issues, but if tried locally the links work.
> > 
> > 
> > The URL checker has a bug. You'll have to merge with the failing workflow until the bug is fixed.
> 
> I've pulled the fix: #35577. Can we wait for the fix to get merged to test the GitHub workflow? Then, we can rebase the branch and see if it works in Github.

Thank you @OnkarRuikar ! I'm happy to wait but it's up to Hamish, IDK if he has a schedule for merging this.

hamishwillee committed:
Update files/en-us/web/api/subtlecrypto/derivebits/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee commented on files/en-us/web/api/subtlecrypto/derivekey/index.md:
```diff
@@ -187,10 +194,277 @@ async function agreeSharedSecretKey() {
+We also set the derived key to be non-extractable, and suitable for encryption and decryption.
+
+The function returns the derived AES-GCM key.
+We use this further down in the code to create shared keys for Bob and Alice.
```
Ah, the "this" is ambiguous. This is supposed to be "We use this _function_ further down ..."

wbamberg replied:
Oh sorry, I understand now. But yes, it is ambiguous.

hamishwillee committed:
Fix ambiguity in the derived key

hamishwillee committed:
Apply suggestions from code review

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee committed:
Purge cyphertext

hamishwillee committed:
Glossary: Key - shared/secret

hamishwillee commented:
Thanks very much for the review (again). I can wait, but I'd like to get this in sooner rather than later. Have pinged Vadim on the other PR.

wbamberg reviewed (approved):
👍 thank you for the updates!

wbamberg committed:
Merge branch 'main' into ff130_x25519_algo
